### ISSUE TITLE
Single Responsibility Principle in Practice And Don't Be Lazy

### ISSUE THREAD

**Issue by tian-li** [tian-li/blog#11]
Yea, another bug to solve while working. This is more like a best practice related bug I guess.

## TL; DR

1. Follow Single Responsibility Principle when writing methods, though it normally applies to classes and modules.
2. Try to only touch things within itself (function, class).

## Background

The application uses [ng-bootstrap](https://github.com/ng-bootstrap/ng-bootstrap) (or ngb for short) for UI components. We have a button can open a modal. The bug is, the modal automatically dismissed when the data in that page is updated.

In case you didn't know, here are some ng-bootstrap modal apis used in this case. Please refer to [their documents](https://ng-bootstrap.github.io/#/components/modal/api) for details, I'm not copying them here.

### APIs in use

* `open`
* `dismiss`
* `dismissAll`
* `beforeDismiss`
* `result`

## Solving Process

My first thought was change detection might somehow trigger ng-bootstrap modal dismiss method since the reject callback of the modal was called when data update came. Luckily ngb has `beforeDismiss` to call before the modal is actually dismissed. So I did various checks in `beforeDismiss` including sync and async using Promise and Observable. Resulting a super complex logic, which my gut feeling told me it shouldn't be. I have to check every dismiss case, making it not future-proof (Not to mentation still had one case couldn't be covered). What if some additional case comes? Add another `if` statement?

Then I suddenly had an idea: maybe the modal was manually dismissed by another method? So I put a break point at the reject callback of the modal, and traced back in the call stack.

And, bingo! `dismissAll`! Here you are! It's blindly dismissing all modals when data is updated.

### Root Cause

In the component that was calling `dismissAll`, it has two places opening a modal. 

The first one opened a modal via a service who was calling ngb's `open` method. The method was called `openModal`, but did a lot of things beyond "open" a modal and didn't return the modal reference.

The second one opened a modal by directly calling ngb's `open` method. Didn't save the modal reference, either.

So in the component, it called `dismissAll` to dismiss whatever modal was open. 

### Solution

With the root cause being known. Just save all the modal reference, and dismiss the modal you want to, instead of calling `dismissAll` blindly.

And problem solved! 🎉 

## Conclusion

The `openModal` method mentioned above did too much things other than just "open" the modal.

The component opened several modals accidentally touched modals opened somewhere else, making it kinda like "impure" if borrowing the concept from "pure function".
